supervisor/shared/web_workflow/web_workflow.c

```diff
@@ -347,8 +347,7 @@ bool supervisor_start_web_workflow(bool reload) {
         #if CIRCUITPY_MDNS
         // Try to start MDNS if the user deinited it.
         if (mdns.base.type != &mdns_server_type ||
-            common_hal_mdns_server_deinited(&mdns) ||
-            reload) { // Always reconstruct on reload, since we don't know if the net changed.
+            common_hal_mdns_server_deinited(&mdns)) {
```

Comment:
This produces a hang on the following scenario:
 - Let wifi autoconnect and web workflow start
 - Disconnect wifi from user code
 - Reload.

It's best left alone.
I tried fixing it in a more elaborate way, but we ultimately can't account for everything.

Reply:
Please give test code for the second step. I'll look into it then. Is the expectation that the reload will reconnect?

Reply:
Reloads do now reconnect as long as creds are set.
Wifi is reenabled too if it was disabled.

Exact steps:
1. Install a `settings.toml` with wifi credentials and api password.
2. Reboot the board.
3. Ensure web-workflow is up.
4. From repl: `import wifi; wifi.radio.enabled=0`
5. Reload with Ctrl+D.
6. ~~Profit.~~ Hardfault.

In any case, the fix is not simple (if it was, I would have done it already). This should be done in a seperate pr.

Reply:
Also, just tested it.. that always was an issue in 8.x...
This actually turned it from a hardfault to a partially broken mdns.

Reply:
Ah and I should probably let you know why it's not an easy fix.

Do keep in mind however, I may very well have misjudged some stuff (due to my lack of understanding of mdns), so do take the following with a grain of salt.

If any network change occurs (even a reconnection to the same network), the mdns object effectively breaks and needs to be recreated before it can be used, or a hardfault occurs when it's used.

On top of that, simply adding:
```c
if (reload && !common_hal_mdns_server_deinited(&mdns)){
    common_hal_mdns_server_deinit(&mdns);
}
```
right after the `|| reload` in question, to deinit the "expired" object, it hardfaults due to `mdns_free();` in `ports/espressif/common-hal/mdns/Server.c:90`.

I have no idea why. Does it need to be deinited before a disconnection?
If it was already deinited, it would just `return`, not explode.

In any case, we would have to know in every reload if the network has changed. Like, even if we were to just record the status in `web_workflow.c`, the user could go, `.enabled=0`, `.connect(...)`, and the web_workflow wouldn't know it needed to remake mdns.

So first we need to be *able* to deinit mdns.
Second we need to know when a mdns reset is needed (perhaps some fancy random connection id we generate in wifi code and can test against?).

Reply:
> I have no idea why. Does it need to be deinited before a disconnection?
> If it was already deinited, it would just `return`, not explode.

Ya, this is my assumption. It looks like the underlying mdns is freed with the IDF but the outer CP object doesn't know. 

https://github.com/adafruit/circuitpython/blob/e39fbf1b26b4fd3b66313e51ccc3db0eba7bd58a/ports/espressif/common-hal/wifi/Radio.c#L94-L105

Reply:
Perhaps we should "trap" any disconnections (due to range, with `.connected`) and `.enabled=0`, and have it reset `STATIC bool mdns_started = false;` in `/home/bill88t/git/circuitpython/ports/espressif/common-hal/mdns/Server.c:39`?
Perhaps also add checks so that if it's `false`, `common_hal_mdns_server_deinited` always returns true.

Reply:
#8418 will fix this issue. Removing the reload here fixes the MDNS info in the meantime.